### PR TITLE
[dy] Add git sync

### PR DESCRIPTION
**Pull request by dy46**
# Summary

Add git sync functionality so that users can sync their data with a remote git repo. I added a `.preferences.yaml` file to store the sync preferences, and will add it to the gitignore so that it doesn't get overwritten when performing a sync. 

Screenshot 2023-02-28 at 9 24 05 AM

# Tests

tested locally, will add unit tests

cc: @wangxiaoyou1993

### CODE REVIEW

File: mage_ai/data_preparation/repo_manager.py [hunk 1]
```diff
@@ -68,6 +68,7 @@ def __init__(self, repo_path: str = None, config_dict: Dict = None):
             self.gcp_cloud_run_config = repo_config.get('gcp_cloud_run_config')
             self.notification_config = repo_config.get('notification_config', dict())
             self.queue_config = repo_config.get('queue_config', dict())
+            self.sync_config = repo_config.get('sync_config', dict())
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/repo_manager.py:71]: Do you put the sync_config in metadata.yaml or preferences.yaml?

File: mage_ai/data_preparation/sync/git_sync.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from mage_ai.data_preparation.sync.base_sync import BaseSync
+from typing import Dict
+import git
+import os
+
+
+class GitSync(BaseSync):
+    def __init__(self, sync_config: Dict):
+        self.remote_repo_link = sync_config.get('remote_repo_link')
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/sync/git_sync.py:9]: I would suggest define a. dataclass with schema for sync_config.

File: mage_ai/api/presenters/SyncPresenter.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+from mage_ai.api.presenters.BasePresenter import BasePresenter
+
+
+class SyncPresenter(BasePresenter):
+    default_attributes = [
+        'type',
+        'remote_repo_link',
```
Comment (wangxiaoyou1993) [inline mage_ai/api/presenters/SyncPresenter.py:7]: nit: alphabetical order

File: mage_ai/data_preparation/preferences.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import yaml
+
+
+class Preferences:
+    def __init__(self, repo_path: str = None, config_dict: Dict = None):
+        self.repo_path = repo_path or get_repo_path()
+
+        try:
+            if not config_dict:
+                metadata_path = os.path.join(self.repo_path, '.preferences.yaml')
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/preferences.py:14]: nit: define a constant for the file name

File: mage_ai/data_preparation/repo_manager.py (hunk 1 above)
Comment (dy46) [reply mage_ai/data_preparation/repo_manager.py:71]: oops let me delete this

**Comment by dy46** [comment]
@wangxiaoyou1993 PTAL

File: mage_ai/data_preparation/sync/__init__.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+@dataclass
+class SyncConfig(BaseConfig):
+    type: SyncType
+    remote_repo_link: str
+    repo_path: str = os.getcwd()
+    branch: str = 'main'
+    sync_on_pipeline_run: bool = False
+
+    def to_dict(self) -> Dict:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/sync/__init__.py:20]: nit: this can be put in the BaseConfig class

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**